pinot-core/src/main/java/org/apache/pinot/core/segment/processing/partitioner/PartitionerFactory.java

```diff
@@ -0,0 +1,82 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.segment.processing.partitioner;
+
+import com.google.common.base.Preconditions;
+
+
+/**
+ * Factory for Partitioner and PartitionFilter
+ */
+public final class PartitionerFactory {
+
+  private PartitionerFactory() {
+
+  }
+
+  public enum PartitionerType {
+    NO_OP, ROW_HASH, COLUMN_VALUE, TRANSFORM_FUNCTION, TABLE_PARTITION_CONFIG
+  }
+
+  /**
+   * Construct a Partitioner using the PartitioningConfig
+   */
+  public static Partitioner getPartitioner(PartitioningConfig config) {
+
+    Partitioner partitioner = null;
+    switch (config.getPartitionerType()) {
+      case NO_OP:
+        partitioner = new NoOpPartitioner();
+        break;
+      case ROW_HASH:
+        Preconditions
+            .checkState(config.getNumPartitions() > 0, "Must provide numPartitions > 0 for ROW_HASH partitioner");
+        partitioner = new RowHashPartitioner(config.getNumPartitions());
+        break;
+      case COLUMN_VALUE:
+        Preconditions.checkState(config.getColumnName() != null, "Must provide columnName for COLUMN_VALUE partitioner");
+        partitioner = new ColumnValuePartitioner(config.getColumnName());
+        break;
+      case TRANSFORM_FUNCTION:
+        Preconditions.checkState(config.getTransformFunction() != null,
+            "Must provide transformFunction for TRANSFORM_FUNCTION partitioner");
+        partitioner = new TransformFunctionPartitioner(config.getTransformFunction());
+        break;
+      case TABLE_PARTITION_CONFIG:
```

Comment:
Since these are all derived segments (merging m segments into n) should we not be using the table's partitioner all the time?

Reply:
Yes. Seunghyun pointed out similar thing. Discussion below.
I'll be changing the partitioner to a 2 step partitioner in the next PR (have put TODO in code and descritpion)